project/npda/general_functions/csv/csv_upload_celery.py

```diff
@@ -0,0 +1,227 @@
+    submission, dataframe: pd.DataFrame, TESTING=False
+):
+    """
+    Process the dataframe and validate and save the patients and visits
+    """
+
+    # Helper functions - these are nested as they are only used in this function. Note that they are duplicated in the celery task as they are not serializable
+    # We can't pass the dataframe to the celery task as it is not serializable
+    # Possibly there is a better way to do this but this is the simplest way to do it
+    def csv_value_to_model_value(model_field, value):
+        if pd.isnull(value) or value == pd.NaT:
+            return None
+
+        if isinstance(value, pd.Timestamp):
+            return value.to_pydatetime().date()
+
+        # Pass Django forms native Python values not numpy ones
+        # https://github.com/rcpch/national-paediatric-diabetes-audit/issues/425
+        return value.item() if isinstance(value, np.generic) else value
+
+    def row_to_dict(row, model, csv_headings):
+        ret = {}
+
+        for entry in csv_headings:
+            if "model" in entry and apps.get_model("npda", entry["model"]) == model:
+                model_field_name = entry["model_field"]
+                model_field_definition = model._meta.get_field(model_field_name)
+
+                csv_value = row[entry["heading"]]
+                model_field_value = csv_value_to_model_value(
+                    model_field_definition, csv_value
+                )
+
+                ret[model_field_name] = model_field_value
+
+        return ret
+
+    # Get the models and set the CSV headings
+    Patient = apps.get_model("npda", "Patient")
+
+    if submission.paediatric_diabetes_unit.pz_code == "PZ248":
+        CSV_HEADINGS = UNIQUE_IDENTIFIER_JERSEY + CSV_HEADING_OBJECTS
+    else:
+        CSV_HEADINGS = UNIQUE_IDENTIFIER_ENGLAND + CSV_HEADING_OBJECTS
+
+    # Group the dataframe by patient
+    # Remember the original row number to help users find where the problem was in the CSV
+    dataframe = dataframe.assign(row_index=np.arange(dataframe.shape[0]))
+
+    # We only one to create one patient per NHS number (or URN if in Jersey) and we can't create their visits if we fail to save the patient model
+    if submission.paediatric_diabetes_unit.pz_code == "PZ248":
+        visits_by_patient = dataframe.groupby(
+            "Unique Reference Number", sort=False, dropna=False
+        )
+    else:
+        visits_by_patient = dataframe.groupby("NHS Number", sort=False, dropna=False)
+
+    # Process each patient and their visits
+    tasks = []
+    for patient_index, patient_group in visits_by_patient:
+        patient_row = patient_group.iloc[0]
+        patient_dict = row_to_dict(patient_row, Patient, csv_headings=CSV_HEADINGS)
+
+        print(f"Submission ID: {submission.id} before saving patient")
+
+        patients_submission_task = save_patient_and_visits_to_submission.s(
+            patient_row_json=patient_row.to_json(date_format="iso"),
```

Comment:
I didn't realise DataFrames supported pickling: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.to_pickle.html.

Could we avoid recleaning the dates in the Celery task? (`convert_invalid_dates_to_none`, `convert_iso_dates`). If Celery doesn't support sending binary data to a task we can base64 encode the pickling.